### PR TITLE
Multi model results class

### PR DESCRIPTION
**Pull request by nv-braf**
Created new Results class (with unit testing) that encapsulates all the ugliness of the list inside list inside tuple of for storing measurement values in checkpoint.

Created a generator function to cycle through all the results and added access functions to replace all the current functionality inside model and metric manager.

I've created a new checkpoint and added instructions on how to create it. This is used in all tests that do not profile, except for the stability tests.

I've also removed one redundant test (L0_export_metric)

### CODE REVIEW

File: model_analyzer/result/results.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        model_config_name: str
+            The model config name 
+            
+        Returns
+        -------
+        bool
+        """
+        return model_config_name in self._results[model_name]
+
+    def is_done(self, model_name):
```
Comment (tgerdesnv) [inline model_analyzer/result/results.py:100]: What is the value of turning this into something that behaves like a generator? Can't we just return the whole list and have the user naturally iterate over it?

File: model_analyzer/result/result_manager.py [hunk 2]
```diff
@@ -350,15 +349,8 @@ def get_model_config_measurements(self, model_config_name):
...
-        if model_name not in results or model_config_name not in results[
-                model_name]:
-            raise TritonModelAnalyzerException(
-                f"Model Config {model_config_name} requested for report step but no results were found. "
-                "Double check the name and ensure that this model config was actually profiled."
-            )
-        else:
-            model_config_data = results[model_name][model_config_name]
-            return model_config_data[0], list(model_config_data[1].values())
+        return results.get_all_model_config_measurements(
```
Comment (tgerdesnv) [inline model_analyzer/result/result_manager.py:352]: This will need a rebase. I fixed a bug 2 lines up.

**Review by tgerdesnv — CHANGES_REQUESTED** [review]

File: model_analyzer/analyzer.py [hunk 3]
```diff
@@ -206,9 +206,8 @@ def _should_profile_multiple_models_concurrently(self):
         return False
 
     def _get_profile_complete_string(self):
-        profiled_model_list = list(
-            self._state_manager.get_state_variable(
-                'ResultManager.results').keys())
+        profiled_model_list = self._state_manager.get_state_variable(
```
Comment (tgerdesnv) [inline model_analyzer/analyzer.py:209]: I was hoping that all of this `_state_manager.get_state_variable('ResultManager.results')` stuff was going to go away, or at least only stored/loaded once. Can we not pull the data into a local instance of your class and operate directly on it?

File: model_analyzer/record/metrics_manager.py [hunk 4]
```diff
@@ -279,13 +280,14 @@ def _get_measurement_if_config_duplicate(self, run_config):
...
             'ResultManager.results')
 
-        # check whether perf config string is a key in result dict
-        if model_name not in results:
+        if not results.contains_model(model_name):
             return False
-        if not self._is_config_in_results(
-                run_config.model_config()._model_config, results[model_name]):
+
+        if not results.contains_model_config(model_name, model_config_name):
```
Comment (tgerdesnv) [inline model_analyzer/record/metrics_manager.py:286]: I would expect that this function can now handle the case above as well (it returns false if the model name doesn't exist, or if the model name exists but the config name doesn't exist in it).

File: model_analyzer/record/metrics_manager.py [hunk 5]
```diff
@@ -279,13 +280,14 @@ def _get_measurement_if_config_duplicate(self, run_config):
...
             return False
-        measurements = results[model_name][model_config_name][1]
+
+        measurements = results.get_model_config_measurements_dict(
```
Comment (tgerdesnv) [inline model_analyzer/record/metrics_manager.py:289]: What is `measurements` here? A list of measurements? A dict?

File: model_analyzer/result/result_manager.py [hunk 6]
```diff
@@ -271,13 +273,9 @@ def add_measurement(self, run_config, measurement):
...
 
-        if model_name not in results:
-            results[model_name] = {}
-        if model_config_name not in results[model_name]:
-            results[model_name][model_config_name] = (model_config, {})
-
-        measurement_key = measurement.perf_config().representation()
-        results[model_name][model_config_name][1][measurement_key] = measurement
+        results.add_measurement(run_config,
+                                measurement.perf_config().representation(),
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply model_analyzer/result/result_manager.py:275]: I don't think this is a good place to put the responsibility. Is the representation only used by the results class? It should either come from the run_config or from the results class.

File: model_analyzer/state/analyzer_state.py [hunk 7]
```diff
@@ -33,24 +34,26 @@ def to_dict(self):
...
-        state._state_dict['ResultManager.results'] = {}
-        for model_name in state_dict['ResultManager.results']:
-            state._state_dict['ResultManager.results'][model_name] = {}
-            for model_config_name in state_dict['ResultManager.results'][
-                    model_name]:
-                model_config_dict, measurements = state_dict[
-                    'ResultManager.results'][model_name][model_config_name]
+        state._state_dict['ResultManager.results'] = Results.from_dict(
+            state_dict['ResultManager.results'])
+        # state._state_dict['ResultManager.results'] = {}
```
Comment (tgerdesnv) [inline model_analyzer/state/analyzer_state.py:39]: can this all be deleted?

**Review by tgerdesnv — CHANGES_REQUESTED** [review]

File: model_analyzer/state/analyzer_state.py (hunk 7 above)
Comment (nv-braf) [reply model_analyzer/state/analyzer_state.py:39]: Deleted

File: model_analyzer/record/metrics_manager.py (hunk 5 above)
Comment (nv-braf) [reply model_analyzer/record/metrics_manager.py:289]: It's a dict (see the results function). This is all going to change when the new measurement class gets implemented, so I didn't feel like putting effort into rewriting this as part of this story.

File: model_analyzer/result/result_manager.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nv-braf) [reply model_analyzer/result/result_manager.py:275]: No, it's also used for duplicate config checking, which is why I didn't think it belonged inside the Results class.

File: model_analyzer/record/metrics_manager.py (hunk 5 above)
Comment (tgerdesnv) [reply model_analyzer/record/metrics_manager.py:286]: nevermind I thought we were changing the return type of the function

File: model_analyzer/analyzer.py (hunk 3 above)
Comment (nv-braf) [reply model_analyzer/analyzer.py:209]: TMA-542 created

File: model_analyzer/config/run/run_config.py [hunk 8]
```diff
@@ -43,6 +43,14 @@ def model_run_configs(self):
         """
         return self._model_run_configs
 
+    def representation(self):
+        """
+        Returns a representation string for the RunConfig that can be used
```
Comment (nv-braf) [inline model_analyzer/config/run/run_config.py:48]: Change the comment to indicate this is a list of MRCs

File: model_analyzer/result/results.py [hunk 9]
```diff
@@ -64,17 +64,18 @@ def from_dict(cls, results_dict):
 
         return results
 
-    def add_measurement(self, model_run_config, key, measurement):
+    def add_measurement(self, model_run_config, measurement):
         """
-        Given a RunConfig and a key, store a measurement
+        Given a RunConfig and a measurement, add the measurement to the
```
Comment (nv-braf) [inline model_analyzer/result/results.py:69]: Given a ModelRunConfig...

File: model_analyzer/result/result_manager.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply model_analyzer/result/result_manager.py]: The changes were made. Resolving

File: model_analyzer/result/results.py (hunk 9 above)
Comment (tgerdesnv) [reply model_analyzer/result/results.py]: fixed

**Review by tgerdesnv — APPROVED** [review]

**Review by tgerdesnv — APPROVED** [review]
